### ISSUE TITLE
could not resolve network endpoint

### ISSUE THREAD

**Issue by marc1uk** [securesocketfunneling/ssf#48]
I'm trying to get started with ssf to tunnel UDP traffic, but i'm unable to connect to the server. I've installed and setup ssf on both client and server side. Certificates were created manually and ca.certs were added from each to the other's /trusted folder. 

I'm trying to route a browser's UDP traffic (a voice conferencing site) via SOCKS proxy to a server outside a firewall. On the server side i'm running 
`./ssfs --port=1223 --verbosity trace -S`
which returns 
```
[2017-10-02 00:34:43][debug] config[ssf]: update config from string:
{
  "ssf": {
    "tls" : {
      "ca_cert_path": "./certs/trusted/ca.crt",
      "cert_path": "./certs/certificate.crt",
      "key_path": "./certs/private.key",
      "key_password": "",
      "dh_path": "./certs/dh4096.pem",
      "cipher_alg": "DHE-RSA-AES256-GCM-SHA384"
    },
    "http_proxy" : {
      "host": "",
      "port": "",
      "credentials": {
        "username": "",
        "password": "",
        "domain": "",
        "reuse_ntlm": true,
        "reuse_nego": true
      }
    },
    "services": {
      "datagram_forwarder": { "enable": true },
      "datagram_listener": {
        "enable": true,
        "gateway_ports": false
      },
      "stream_forwarder": { "enable": true },
      "stream_listener": {
        "enable": true,
        "gateway_ports": false
      },
      "file_copy": { "enable": false },
      "shell": {
        "enable": false,
        "path": "/bin/bash",
        "args": ""
      },
      "socks": { "enable": true }
    }
  }
}

[2017-10-02 00:34:43][info] config[ssf]: loading file <config.json>
[2017-10-02 00:34:43][info] config[tls]: CA cert path: <./certs/trusted/ca.crt>
[2017-10-02 00:34:43][info] config[tls]: cert path: <./certs/certificate.crt>
[2017-10-02 00:34:43][info] config[tls]: key path: <./certs/private.key>
[2017-10-02 00:34:43][info] config[tls]: key password: <PASSWORD_REMOVED>
[2017-10-02 00:34:43][info] config[tls]: dh path: <./certs/dh4096.pem>
[2017-10-02 00:34:43][info] config[tls]: cipher suite: <DHE-RSA-AES256-GCM-SHA384>
[2017-10-02 00:34:43][info] config[HTTP proxy]: <None>
[2017-10-02 00:34:43][info] status[microservices] datagram_forwarder: On
[2017-10-02 00:34:43][info] status[microservices] datagram_listener: On
[2017-10-02 00:34:43][info] status[microservices] stream_forwarder: On
[2017-10-02 00:34:43][info] status[microservices] stream_listener: On
[2017-10-02 00:34:43][info] status[microservices] file_copy: Off 
[2017-10-02 00:34:43][info] status[microservices] shell: Off
[2017-10-02 00:34:43][info] status[microservices] socks: On
[2017-10-02 00:34:43][info] server: listening on <*:1223>
[2017-10-02 00:34:43][debug] async engine: starting
[2017-10-02 00:34:43][info] server: running (Ctrl + C to stop)   
```

On the client side i'm then running 
`./ssfc --port=1223 -D8888 USER@SERVER_IP -v trace -S`
Which returns:
```
[2017-10-01 18:34:38][debug] config[ssf]: update config from string:
{
  "ssf": {
    "tls" : {
      "ca_cert_path": "./certs/trusted/ca.crt",
      "cert_path": "./certs/certificate.crt",
      "key_path": "./certs/private.key",
      "key_password": "",
      "dh_path": "./certs/dh4096.pem",
      "cipher_alg": "DHE-RSA-AES256-GCM-SHA384"
    },
    "http_proxy" : {
      "host": "",
      "port": "",
      "credentials": {
        "username": "",
        "password": "",
        "domain": "",
        "reuse_ntlm": true,
        "reuse_nego": true
      }
    },
    "services": {
      "datagram_forwarder": { "enable": true },
      "datagram_listener": {
        "enable": true,
        "gateway_ports": false
      },
      "stream_forwarder": { "enable": true },
      "stream_listener": {
        "enable": true,
        "gateway_ports": false
      },
      "file_copy": { "enable": false },
      "shell": {
        "enable": false,
        "path": "/bin/bash",
        "args": ""
      },
      "socks": { "enable": true }
    }
  }
}

[2017-10-01 18:34:38][info] config[ssf]: loading file <config.json>
[2017-10-01 18:34:38][info] config[tls]: CA cert path: <./certs/trusted/ca.crt>
[2017-10-01 18:34:38][info] config[tls]: cert path: <./certs/certificate.crt>
[2017-10-01 18:34:38][info] config[tls]: key path: <./certs/private.key>
[2017-10-01 18:34:38][info] config[tls]: key password: <PASSWORD_REMOVED>
[2017-10-01 18:34:38][info] config[tls]: dh path: <./certs/dh4096.pem>
[2017-10-01 18:34:38][info] config[tls]: cipher suite: <DHE-RSA-AES256-GCM-SHA384>
[2017-10-01 18:34:38][info] config[HTTP proxy]: <None>
[2017-10-01 18:34:38][info] status[microservices] datagram_forwarder: On
[2017-10-01 18:34:38][info] status[microservices] datagram_listener: On
[2017-10-01 18:34:38][info] status[microservices] stream_forwarder: On
[2017-10-01 18:34:38][info] status[microservices] stream_listener: On
[2017-10-01 18:34:38][info] status[microservices] file_copy: Off
[2017-10-01 18:34:38][info] status[microservices] shell: Off
[2017-10-01 18:34:38][info] status[microservices] socks: On
[2017-10-01 18:34:38][info] config[circuit]: <None>
[2017-10-01 18:34:38][error] client: could not resolve network endpoint
[2017-10-01 18:34:38][error] client: error happened when running client: Host not found (authoritative)
[2017-10-01 18:34:38][debug] async engine: stopping
```

Can anyone suggest why the client isn't able to connect? 
I've commonly used a normal SSH tunnel to route data, but of course open ssh doesn't work with UDP. As a hackey trial of the ssf application setup, I tried setting up a ssh port forward using 
`ssh -p 1222 -L1223:192.168.1.83:1223 USER@SERVER_IP`
which should forward packets sent to my local port 1223 to the server port 1223 (where my ssf server is listening), then in parallel setting up a ssf socks proxy, pretending my ssf server is in fact local port 1223. 
`./ssfc -p 1223 -D8888 127.0.0.1`
Setting my browser socks proxy to port 8888, this did indeed work - at least for TCP traffic - but did not work as a workaround for UDP traffic. 

Anyone got any pointers? 
The SSF website says 'send us your application logs', but i dunno where those are. 
SSF 2.2.0, Boost 1.61.0 and OpenSSL 1.0.2.8

**Comment by olachere** [comment]
Hi @marc1uk,

There are two problems with what you are trying to achieve here:

- First, SSF does not provide user authentication, and it seems you are trying to specify a username on the command line (USER@SERVER_IP). The error message you see just means that SSF cannot perform name resolution on the provided hostname. Just drop "USER@" in front of the hostname and you should be good to go. Note that SSL automatically authenticates the client to the server using the certificates you produced.

- The SOCKS implementation in SSF does not currently support UDP traffic redirection. It only provides static UDP port forwarding (-U and -V options). It seems that this feature request is already tracked by issue #47

**Comment by marc1uk** [comment]
Hi olachere,
Thanks for pointing that out, removing the USER@ part worked perfectly - the connection was made and socks proxy worked as expected. :+1: 
I was a little surprised that ssf seems distinctly slower than a standard ssh port dynamic port forward - not substantially, but noticably. 
Anyway, the crucial point for me here is your second one - the website wasn't clear whether UDP forwarding was supported as part of a SOCKS proxy, so thanks for clarifying that. Indeed, although browsing now works via SSF SOCKS, the voice chat doesn't. 
Thanks for the pointer to the issue, I'll follow that and hope for the feature in the future.

Cheers! :)
